### PR TITLE
Simplify fence functions in the Threads backend

### PR DESCRIPTION
**Pull request by Rombur**
Rework fencing in the Threads backend. Currently, we use an `enum` to call the right profiling event in `internal_fence`. The profiling event calls a lambda function defined inside `internal_fence`. This PR removes the enum and move the profiling events to the `fence` and `impl_static_fence` functions. The profiling events now call `internal_fence` which only contains the fencing logic.
This PR also changes the profiling message in `ThreadsInternal::fence()` from `Kokkos::ThreadsInternal::fence: Unnamed Static Fence` to `Kokkos::ThreadsInternal::fence: Unnamed Instance Fence`

### CODE REVIEW

**Review by masterleinad — APPROVED** [review]

**Review by dalg24 — COMMENTED** [review]
Did you look at tools output?

File: core/src/Threads/Kokkos_Threads_Instance.cpp [hunk 1]
```diff
@@ -253,46 +253,29 @@ int ThreadsInternal::in_parallel() {
   return s_current_function && (&s_threads_process != s_current_function_arg) &&
          (s_threads_process.m_pool_base || !is_process());
 }
-void ThreadsInternal::fence() { internal_fence(Impl::fence_is_static::yes); }
-void ThreadsInternal::fence(const std::string &name) {
-  internal_fence(name, Impl::fence_is_static::yes);
+void ThreadsInternal::fence() {
+  fence("Kokkos::ThreadsInternal::fence : Unnamed Instance Fence");
```
[the remark this answers is not in the document]
Comment (Rombur) [reply core/src/Threads/Kokkos_Threads_Instance.cpp:257]: Right but the message was wrong before, it is not a static fence. In truth, there is currently no difference between a static fence and instance fence in the Threads backend.

[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/Threads/Kokkos_Threads_Instance.cpp]: Please make sure the description capture that this PR corrected that

**Review by dalg24 — APPROVED** [review]

**Comment by dalg24** [comment]
Ignored timeout in one HIP build

**Pull request merged**